pontoon/sync/core/__init__.py

```diff
@@ -0,0 +1,96 @@
+import logging
+
+from notifications.signals import notify
+
+from django.utils import timezone
+
+from pontoon.base.models import ChangedEntityLocale, Locale, Project, User
+from pontoon.pretranslation.tasks import pretranslate
+from pontoon.sync.core.checkout import checkout_repos
+from pontoon.sync.core.entities import sync_entities_from_repo
+from pontoon.sync.core.paths import find_paths
+from pontoon.sync.core.translations_from_repo import sync_translations_from_repo
+from pontoon.sync.core.translations_to_repo import sync_translations_to_repo
+from pontoon.sync.models import ProjectSyncLog, RepositorySyncLog, SyncLog
+
+
+log = logging.getLogger(__name__)
+
+
+def sync_project(
+    project: Project,
+    sync_log: SyncLog,
+    *,
+    pull: bool = True,
+    commit: bool = True,
+    force: bool = False,
+):
+    # Mark "now" at the start of sync to avoid messing with
+    # translations submitted during sync.
+    now = timezone.now()
+
+    log_prefix = f"[{project.slug}]"
+    log.info(f"{log_prefix} Sync start")
+    project_sync_log = ProjectSyncLog.objects.create(
+        sync_log=sync_log, project=project, start_time=now
+    )
+
+    try:
+        checkouts = checkout_repos(project, pull, force)
+        paths = find_paths(project, checkouts)
+    except Exception as e:
+        log.error(f"{log_prefix} {e}")
+        project_sync_log.skip()
+        raise e
+
+    locale_map: dict[str, Locale] = {
+        lc.code: lc for lc in project.locales.order_by("code")
+    }
+    paths.locales = list(locale_map.keys())
+    added_entities_count, changed_paths, removed_paths = sync_entities_from_repo(
+        project, locale_map, checkouts.source, paths, now
+    )
+
+    repo_sync_log = RepositorySyncLog.objects.create(
+        project_sync_log=project_sync_log,
+        repository=checkouts.target.repo,
+        start_time=timezone.now(),
+    )
+
+    db_changes = ChangedEntityLocale.objects.filter(
+        entity__resource__project=project, when__lte=now
+    ).select_related("entity__resource", "locale")
+    sync_translations_from_repo(project, locale_map, checkouts, paths, db_changes, now)
+    if added_entities_count > 0:
+        notify_users(project, added_entities_count)
+    sync_translations_to_repo(
+        project,
+        commit,
+        locale_map,
+        checkouts,
+        paths,
+        db_changes,
+        changed_paths,
+        removed_paths,
+        now,
+    )
+    db_changes.delete()
+    # have_repos_changed = any(co.commit is None or co.commit != co.prev_commit for co in checkouts)
+    # pulled_revisions = {co.locale_code or "single_locale": co.commit for co in checkouts}
+    log.info(f"{log_prefix} Sync done")
+    repo_sync_log.end()
+
+    if project.pretranslation_enabled and changed_paths:
+        # Pretranslate changed and added resources for all locales
+        pretranslate(project, changed_paths)
```

Comment:
The new sync is retaining slightly different information than before, so it makes more sense to trigger the pretranslation for all entities in changed resources. The pretranslation will then filter out the entities and locales for which pretranslation is not needed or wanted.